arwen/contexts/storage.go

```diff
@@ -379,14 +384,23 @@ func (context *storageContext) computeGasForKey(key []byte, usedCache bool) uint
 }
 
 // UseGasForStorageLoad - single spot of gas consumption for storage load
-func (context *storageContext) UseGasForStorageLoad(tracedFunctionName string, loadCost uint64, usedCache bool) {
+func (context *storageContext) UseGasForStorageLoad(tracedFunctionName string, trieDepth int64, staticGasCost uint64, usedCache bool) error {
+	var blockchainLoadCost uint64
+	var err error
+
 	metering := context.host.Metering()
 	enableEpochsHandler := context.host.EnableEpochsHandler()
 	if enableEpochsHandler.IsStorageAPICostOptimizationFlagEnabled() && usedCache {
-		loadCost = metering.GasSchedule().ElrondAPICost.CachedStorageLoad
+		blockchainLoadCost = metering.GasSchedule().ElrondAPICost.CachedStorageLoad
+	} else {
+		blockchainLoadCost, err = context.GetStorageLoadCost(trieDepth, staticGasCost)
+		if err != nil {
+			return err
+		}
 	}
 
-	metering.UseGasAndAddTracedGas(tracedFunctionName, loadCost)
+	metering.UseGasAndAddTracedGas(tracedFunctionName, blockchainLoadCost)
```

Comment:
Now that `storage.UseGasForStorageLoad()` is returning an `error`, consider using `metering.UseGasBounded()` instead of `metering.UseGasAndAddTracedGas()`, because `metering.UseGasBounded()` returns an early error if the contract doesn't have enough gas for the storage load.

Even better, consider creating a new method `metering.UseGasBoundedAndAddTracedGas()`, which is similar to `metering.UseGasAndAddTracedGas()` but differs by calling `UseGasBounded()` within.

This way you can return the error from `metering.UseGasBounded()` directly instead of `nil`.

Reply:
Added new method.